### PR TITLE
Fix NPM packages name validation

### LINKED ISSUE #26551: NPM Package Registry returns HTTP400 for packages with single-character names
### Description

Steps to reproduce the issue:
1. Create new token or find existing token with privileges to publish packages.
1. Install NPM locally or use container.
1. Run `npm i q` in environment from previous step
It'll download `https://www.npmjs.com/package/q/v/1.5.1` as your dependency and put it in `./node_modules/q/` directory.
1. Run
    ```
    npm config set registry http://{gitea_url}:3000/api/packages/{username}/npm/
    npm config set -- '//{gitea_url}:3000/api/packages/{username}/npm/:_authToken' "{token}"
    ```
1. Run `npm publish ./node_modules/q/` to publish this package to Gitea.
1. Error from NPM
   ```
   npm ERR! code E400
   npm ERR! 400 Bad Request - PUT http://{gitea_url}:3000/api/packages/{username}/npm/q - package name is invalid
   ```
Error from Gitea log
```
2023/08/16 23:11:16 ...ges/helper/helper.go:34:LogAndProcessError() [D] package name is invalid
2023/08/16 23:11:16 ...eb/routing/logger.go:102:func1() [I] router: completed PUT /api/packages/{username}/npm/q for 172.18.0.1:52196, 400 Bad Request in 19.4ms @ npm/npm.go:154(npm.UploadPackage)
```
   

### Gitea Version

3b129aa

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

2.40.1

### Operating System

Ubuntu 22.04

### How are you running Gitea?

Gitea running in docker.

Docker version:
```
vagrant@ubuntu:~$ docker version
Client:
 Version:           20.10.25
 API version:       1.41
 Go version:        go1.18.1
 Git commit:        20.10.25-0ubuntu1~22.04.1
 Built:             Fri Jul 14 21:58:09 2023
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.25
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.1
  Git commit:       20.10.25-0ubuntu1~22.04.1
  Built:            Thu Jun 29 21:21:05 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.2
  GitCommit:
 runc:
  Version:          1.1.7-0ubuntu1~22.04.1
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:
```

With docker-compose version:
```
vagrant@ubuntu:~$ docker-compose version
docker-compose version 1.29.2, build unknown
docker-py version: 5.0.3
CPython version: 3.10.12
OpenSSL version: OpenSSL 3.0.2 15 Mar 2022
```

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by TimberBro**
- Added new tests to cover corner cases
- Replace existing regex with new one
Closes #26551 

---
As @silverwind suggested, I started from [validate-npm-package-name](https://github.com/npm/validate-npm-package-name), but found this solution too complicated.
Then I tried to fix existing regex myself, but thought, that exclude all restricted symbols is harder, than set only allowed symbols.
Then I search a bit more and found [package-name-regex](https://github.com/dword-design/package-name-regex) and regex from it works for all new test cases.

Let me know, if more information or help with this PR is needed.

### CODE REVIEW

File: modules/packages/npm/creator.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ var (
 	ErrInvalidIntegrity = util.NewInvalidArgumentErrorf("failed to validate integrity")
 )
 
-var nameMatch = regexp.MustCompile(`\A((@[^\s\/~'!\(\)\*]+?)[\/])?([^_.][^\s\/~'!\(\)\*]+)\z`)
+var nameMatch = regexp.MustCompile(`^(@[a-z0-9-~][a-z0-9-._~]*/)?[a-z0-9-~][a-z0-9-._~]*$`)
```
Comment (delvh) [inline modules/packages/npm/creator.go]: > package name should not contain any of the following characters: `~)('!*`
```suggestion
var nameMatch = regexp.MustCompile(`^(@[a-z0-9-][a-z0-9-._]*/)?[a-z0-9-][a-z0-9-._]*$`)
```

File: modules/packages/npm/creator_test.go [hunk 2]
```diff
@@ -67,6 +67,16 @@ func TestParsePackage(t *testing.T) {
...
+		test(t, "^test")
+		test(t, "te^st")
+		test(t, "te|st")
+		test(t, "te)(st")
+		test(t, "te'st")
+		test(t, "te!st")
+		test(t, "te*st")
 		test(t, "invalid/scope")
 		test(t, "@invalid/_name")
 		test(t, "@invalid/.name")
```
Comment (delvh) [inline modules/packages/npm/creator_test.go:83]: I think we should also test for
> package name cannot be the same as a node.js/io.js core module nor a reserved/blacklisted name. For example, the following names are invalid:
    http
    stream
    node_modules
    favicon.ico

Comment (silverwind) [reply modules/packages/npm/creator_test.go:83]: Could keep a map of reserved names. Would be cleaner then extending the regex.

Comment (delvh) [reply modules/packages/npm/creator_test.go:83]: One doesn't exclude the other?
The test method does not test the regex directly, but the upload.

File: modules/packages/npm/creator.go (hunk 1 above)
Comment (TimberBro) [reply modules/packages/npm/creator.go]: Thanks! Added test for that case and applied suggested solution.

File: modules/packages/npm/creator_test.go (hunk 2 above)
Comment (TimberBro) [reply modules/packages/npm/creator_test.go:83]: Can we merge this PR without reserved names inspection feature?
Can I create separated issue for that, and I think, I'm able to implement it, but need some time to dive into codebase.

**Review by delvh — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Can do more later.

**Pull request merged**